docs/site/pages/concepts/controllers.mdx

```diff
@@ -8,4 +8,245 @@ Head to the configuration reference section for [Controllers](/configuration/con
+By default, Flipt ships with a basic controller known as the `template` controller and an extensible controller called the `wasm` controller.
+
+<Callout type="info" emoji="ℹ️">
+You will learn in the next section on [bindings](/concepts/bindings) how to configure which controllers handle which resource definitions.
+</Callout>
+
+Controllers are the most important extension point in Flipt.
+They allow operators to create custom transformation logic between API resources and their resulting representation in the repository.
+Meaning there is no strict requirement that these two concepts need to have a 1:1 mapping (though the `template` controller does support this kind of logic out of the box).
+
+## Lifecycle of a Cup Resource Request
+
+Cup automatically exposes four verbs for each configured resource:
+
+- Get
+- List
+- Put
+- Delete
+
+The first two are read-only operations and the second two produce written changes to state.
+Reads simply source the current state from Git for a particular revision through the means of a controller to handle parsing source and serializing responses.
+Writes also go through to Git via the means of a controller, however, they also result in changes being presented as a pull request.
+
+The following two sections attempts to explain these processes in a little more depth.
+
+### Read Requests
+
+```mermaid
+sequenceDiagram
+    participant A as User
+    participant S as API Server
+    participant C as Controller
+    participant G as Git
+    A ->>+ S: GET /apis/g/v/k/...
+    S ->>+ G: checkout
+    S ->>+ C: Get(g, v, k)
+    C -->> G: read()
+    G -->>- C: []byte
+    C -->>- S: return Resource{}
+    S -->>- A: 200 {"apiVersion": ...}
+```
+
+The API server is the first stop for all operations on resources in Cup.
+It takes care of collecting up the relevant state from Git and delegating onto the relevant **controller**.
+Controllers have no concept of Git, they work purely over a filesystem abstraction.
+It is up to the API layer to build the expected implementation of the filesystem for a particular revision.
+This filesystems root will be the root directory of the target Git repository being sourced from. 
+
+### Write Requests
+
+```mermaid
+sequenceDiagram
+    participant A as User
+    participant S as API Server
+    participant C as Controller
+    participant G as Git
+    participant SCM as SCM
+    A ->>+ S: PUT /apis/g/v/k/...
+    S ->>+ G: checkout new branch
+    S ->>+ C: PUT(Resource{})
+    C ->> G: write([]byte)
+    C ->> G: add, commit and push
+    G -->>- C: 
+    C -->>- S: return
+    S ->> SCM: OpenPR(branch, title)
+    SCM -->> S: PR{}
+    S -->>- A: 202 Accepted
+```
+
+Write requests on the other hand are a little more complicated.
+In this scenario, the filesystem abstraction passed to a Controller can handle writes.
+The API server which builds the implementation of the filesystem, will do so over a worktree on a new branch.
```

Comment:
```suggestion
In Git's case, the API server which builds the implementation of the filesystem, will do so over a worktree on a new branch.
```
You can take it or leave it.

I had to piece it a bit together. Are there any other FS abstractions that cup supports?

Reply:
Yeah it supports local too. Like with Flipt.